### PR TITLE
virtualization: properly restore cold/warm states, cache, and scry stack when catching an error

### PR DESCRIPTION
**Pull request by eamsden**
Fixed a bug caught by @ashelkovnykov 

@ashelkovnykov please confirm my understanding. In particular:

- [ ] `jet_mute` and `jet_mure` had the same bug as `mink`
- [ ] `jet_mute` and `jet_mure` additionally did not properly reset the scry stack in the context when catching an error.
- [ ] (not changed yet) `mink` further does not properly reset the scry stack in the context when catching an error

### CODE REVIEW

**Comment by ashelkovnykov** [comment]
* `jet_mure` and `jet_mute` were also not saving `Cold`/`Warm` state correctly
* `jet_mure` and `jet_mute` did not correctly reset the scry stack on deterministic error & blocked scry
* `mink` **does** reset the scry stack correctly in all cases (via `jet_mink`)

File: rust/ares/src/jets/nock.rs [hunk 1]
```diff
@@ -90,21 +85,26 @@ pub fn jet_mure(context: &mut Context, subject: Noun) -> Result {
...
-                context.restore(saved);
-                context.scry_stack = context.scry_stack.as_cell()?.tail();
+            Error::ScryBlocked(_) => {
+                context.scry_stack = snapshot.scry_stack;
+                Ok(D(0))
+            }
+            Error::Deterministic(_) => {
+                context.cold = snapshot.cold;
+                context.warm = snapshot.warm;
+                context.scry_stack = snapshot.scry_stack;
```
Comment (eamsden) [inline rust/ares/src/jets/nock.rs:107]: Do we not need to reset the cache here?

File: rust/ares/src/jets/nock.rs [hunk 2]
```diff
@@ -116,29 +116,29 @@ pub fn jet_mute(context: &mut Context, subject: Noun) -> Result {
...
             Ok(T(&mut context.stack, &[YES, res]))
         }
         Err(error) => match error {
             // Since we are using the pass-through scry handler, we know for a fact that a scry
             // crash must have come from a senior virtualization context.
             Error::NonDeterministic(_) | Error::ScryCrashed(_) => Err(JetErr::Fail(error)),
             Error::ScryBlocked(path) => {
-                context.scry_stack = context.scry_stack.as_cell()?.tail();
-                context.restore(saved);
+                context.scry_stack = snapshot.scry_stack;
```
Comment (eamsden) [inline rust/ares/src/jets/nock.rs:132]: Do we not need to restore `cold`/`warm`/`cache` before `smyt`()? Won't it be running with invalidated versions of those pointers?

File: rust/ares/src/jets/nock.rs [hunk 3]
```diff
@@ -116,29 +116,29 @@ pub fn jet_mute(context: &mut Context, subject: Noun) -> Result {
...
                 //      return [[%leaf "mute.hunk"] ~] if not
                 let bon = util::smyt(&mut context.stack, path)?;
                 Ok(T(&mut context.stack, &[NO, bon, D(0)]))
             }
             Error::Deterministic(trace) => {
-                context.scry_stack = context.scry_stack.as_cell()?.tail();
-                context.restore(saved);
+                context.cold = snapshot.cold;
+                context.warm = snapshot.warm;
+                context.scry_stack = snapshot.scry_stack;
```
Comment (eamsden) [inline rust/ares/src/jets/nock.rs:141]: As above: do we not restore the old `cache` on failure?

File: rust/ares/src/serf.rs [hunk 4]
```diff
@@ -339,6 +341,8 @@ fn soft(context: &mut Context, ovo: Noun, trace_name: Option<String>) -> Result<
         Ok(res) => Ok(res),
         Err(error) => match error {
             Error::Deterministic(trace) | Error::NonDeterministic(trace) => {
+                context.nock_context.cold = cold_snapshot;
+                context.nock_context.warm = warm_snapshot;
```
Comment (eamsden) [inline rust/ares/src/serf.rs:345]: `cache` should possibly be reset to a new Hamt here, if this method is only used in serf. Otherwise it should be saved and restored, right?

**Review by eamsden — COMMENTED** [review]
Mostly just: "shouldn't we restore the saved cache after failure?" in a bunch of different places. I might not be following something...

**Comment by eamsden** [comment]
Notes from discussion with @ashelkovnykov this morning

- `interpret()` should restore everything except the scry stack on error (since mink needs to check the scry stack as of the error to see if it should re-throw scry-crashed or convert to another error
- All virtualization functions should call `util::mink`

**Comment by ashelkovnykov** [comment]
Follow-up to @eamsden 's comments above: I tried to elide several calls and unnecessary restorations in `jet_mink`, `jet_mure`, `jet_mute`, but it's making more problems than it's solving. It's resulting in the logic getting slightly more spread out, but also - much worse than that - it's making the logic of when we do/don't have a valid cold state / warm state / cache / scry stack hard to follow.

Therefore, we're going to put all of the logic (except scry stack) into `interpret` itself, and then force all possible virtualization through `util::mink`. Then we can verify all of the `Context` is in the correct state by checking only three locations: `jets::nock::util::mink`, `interpreter::interpret`, and `interpreter::exit`.

File: rust/ares/src/jets/nock.rs (hunk 1 above)
Comment (ashelkovnykov) [reply rust/ares/src/jets/nock.rs:107]: The logic is that the error is going to be propagated up to either a higher level `+mink` which will catch it and reset the cache, or to the root serf process - meaning that the cache is going to be reset anyways, as the event processing is over.

See comments below for planned changes.

File: rust/ares/src/jets/nock.rs (hunk 2 above)
Comment (ashelkovnykov) [reply rust/ares/src/jets/nock.rs:132]: `+smyt` just parses a `$path` to a `$tank`

File: rust/ares/src/jets/nock.rs (hunk 3 above)
Comment (ashelkovnykov) [reply rust/ares/src/jets/nock.rs:141]: Same as above; see comments below for planned changes.

File: rust/ares/src/serf.rs (hunk 4 above)
Comment (ashelkovnykov) [reply rust/ares/src/serf.rs:345]: The cache should not be saved / restored here, since it's empty before going into this call and either an error or a success mean that it's about to be invalidated (since it's not a persistent cache). However, you're right that we need to reset it explicitly in the error case: the call to `goof` should have an empty cache.

See comments below for planned changes.

**Review by eamsden — COMMENTED** [review]
This looks good to me now, but it needs to be tested booting a fakezod and someone not me or @ashelkovnykov needs to review it.

**Comment by ashelkovnykov** [comment]
@eamsden Currently failing on fakezod boot
